tests/unit/s2n_serialize_connection_test.c

```diff
@@ -0,0 +1,476 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+#include "tls/s2n_config.h"
+#include "tls/s2n_connection.h"
+
+#define S2N_SERIALIZED_CONN_TLS13_SHA256_SIZE 126
+
+#define TEST_SEQUENCE_NUM 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+#define TEST_TLS13_SECRET 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, \
+                          0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, \
+                          0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key *chain_and_key = NULL,
+            s2n_cert_chain_and_key_ptr_free);
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    DEFER_CLEANUP(struct s2n_config *tls12_config = s2n_config_new(), s2n_config_ptr_free);
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(tls12_config, chain_and_key));
+    EXPECT_SUCCESS(s2n_config_disable_x509_verification(tls12_config));
+    EXPECT_SUCCESS(s2n_config_set_serialized_connection_version(tls12_config, S2N_SERIALIZED_CONN_V1));
+
+    DEFER_CLEANUP(struct s2n_config *tls13_config = s2n_config_new(), s2n_config_ptr_free);
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(tls13_config, chain_and_key));
+    EXPECT_SUCCESS(s2n_config_disable_x509_verification(tls13_config));
+    EXPECT_SUCCESS(s2n_config_set_serialized_connection_version(tls13_config, S2N_SERIALIZED_CONN_V1));
+    /* Security policy that can negotiation TLS13 and has aes_128_gcm_sha256 as its preferred cipher suite */
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(tls13_config, "20210825"));
+
+    /* s2n_connection_serialization_length */
+    {
+        /* Safety */
+        {
+            struct s2n_connection conn = { 0 };
+            uint32_t length = 0;
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_serialization_length(&conn, NULL), S2N_ERR_NULL);
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_serialization_length(NULL, &length), S2N_ERR_NULL);
+        };
+
+        /* Length is correct for all possible cipher suites in TLS1.3 */
+        {
+            DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(conn);
+
+            conn->config->serialized_connection_version = S2N_SERIALIZED_CONN_V1;
+            conn->actual_protocol_version = S2N_TLS13;
+            struct s2n_cipher_suite *tls13_cipher_suites[] = {
+                &s2n_tls13_aes_128_gcm_sha256,
+                &s2n_tls13_aes_256_gcm_sha384,
+                &s2n_tls13_chacha20_poly1305_sha256,
+            };
```

Comment:
This won't get updated. Probably best to use cipher_preferences_test_all_tls13, or maybe even better, security_policy_test_all_tls13->cipher_preferences.